Update nusb to 0.2

thenewwazoo opened the pull request:
Full disclosure right up front: I wrote this PR with LLM tooling assistance, but I am responsible for its content.

This PR updates the nusb dependency to version 0.2. The commit message has more flavor, but it's mostly a mechanical change that relies heavily on the blocking API so as to avoid deviation. Reviewers will likely want to pay the most attention to the updates in `InterfaceExt`.

I have tested this change with a Raspberry Pi Pico Probe, flashing real firmware onto a real device on a macOS host. I also have a JLink probe and some STLinkv2-equipped devices (some Nucleos), and would be happy to test those too (smoke-test? something more?). As this is my first time contributing to this codebase, I'm not quite sure what to offer, but understand that PRs which touch nearly every supported probe might be risky.

I've also tried to meet the contribution guidelines, but please let me know if I've erred.

## Example test:

```
     Running `/Users/bmatt/Development/probe-rs/target/debug/smoke_tester test --chip rp235x --probe '2e8a:000c'`
[1/1](RP235x) - Starting Test
[1/1](RP235x) - Probe: "CMSIS-DAP"
[1/1](RP235x) - Chip:  "RP235x"
[1/1](RP235x) - Core 0: Armv8m
[1/1](RP235x) - Halting core..
[1/1](RP235x) - Test [1] - Testing stepping on core 0...
[1/1](RP235x) - Test [1] - R0 value ok!
[1/1](RP235x) - Test [1] - Core halted at 0x20000002, now trying to run...
[1/1](RP235x) - Test [1] - Core halted again!
[1/1](RP235x) - Test [1] - Core halted at 0x20000006, now trying to run...
[1/1](RP235x) - Test [1] - Run core again, with pc = 0x20000008
[1/1](RP235x) - Test [1] - Test passed in 37 ms.
[1/1](RP235x) - Test [2] - Testing HW breakpoints
[1/1](RP235x) - Test [2] - Testing region: <unnamed region> (0x10000000 - 0x14000000)
[1/1](RP235x) - Test [2] - 8 breakpoints supported
[1/1](RP235x) - Test [2] - Test passed in 65 ms.
[1/1](RP235x) - Test [3] - Testing region: <unnamed region> (0x20000000 - 0x20082000)
[1/1](RP235x) - Test [3] - Test - RAM Start 32
[1/1](RP235x) - Test [3] - Test - RAM End 32
[1/1](RP235x) - Test [3] - Test - RAM Start 8
[1/1](RP235x) - Test [3] - Test - RAM 8 Unaligned
[1/1](RP235x) - Test [3] - Test - RAM End 8
[1/1](RP235x) - Test [3] - Testing: write and read at address 0x20000000: 1 byte at RAM start
[1/1](RP235x) - Test [3] - Testing: write and read at address 0x20000000: 4 bytes at RAM start
[1/1](RP235x) - Test [3] - Testing: write and read at address 0x20081FFC: 4 bytes at RAM end
[1/1](RP235x) - Test [3] - Test passed in 3 ms.
[1/1](RP235x) - Test [4] - Testing register write...
[1/1](RP235x) - Test [4] - Test passed in 26 ms.
[1/1](RP235x) - Test [5] - Testing register read...
[1/1](RP235x) - Test [5] - Test passed in 16 ms.
[1/1](RP235x) - Test [6] - Missing resource for test: No flash test binary specified
[1/1](RP235x) - Tests Passed
[DONE] - All DUTs passed.

Test summary:
 [RP235x] passed
All tests passed.
```

thenewwazoo committed:
Update nusb to 0.2.0

This change updates the nusb dependency to 0.2.0. There were a number of
[significant changes] in that version. This change relies heavily on
using nusb in its "blocking" mode, in accordance with prior behavior.
The most significant change was to the internal logic in `InterfaceExt`,
with the bulk read/write helpers getting a non-trivial rewrite
(Buffer/Endpoint usage with explicit timeouts and cancellations).

This change is otherwise largely mechanical and the public API of the
crate was not changed.

[significant changes]: https://github.com/kevinmehall/nusb/releases/tag/v0.2.0

thenewwazoo committed:
Fix clippy lints

Clippy is cool, I should remember to use it _before_ I drop my PR.

thenewwazoo committed:
Fix mistaken fix

I didn't read clippy's suggestion carefully enough, and broke it.

thenewwazoo committed:
Prefer tracing over env_logger

Because the probe-rs library uses the tracing crate, setting RUST_LOG
doesn't cause the probe-rs library to emit logs. This change moves the
smoke tester to tracing_subscriber, so probe-rs logs can be gathered.

Example output before:
```
$ RUST_LOG=debug RUST_BACKTRACE=1 cargo run -- test --chip STM32G031K8Tx --probe 0483:374B | head
   Compiling probe-rs v0.29.0 (/Users/bmatt/Development/probe-rs/probe-rs)
   Compiling smoke_tester v0.29.0 (/Users/bmatt/Development/probe-rs/smoke-tester)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 3.08s
     Running `/Users/bmatt/Development/probe-rs/target/debug/smoke_tester test --chip STM32G031K8Tx --probe '0483:374B'`
[1/1](STM32G031K8Tx) - Starting Test
[2025-09-19T19:16:24Z DEBUG nusb::platform::macos_iokit::enumeration] Probing device 100536e5b
```

and after:
```
$ RUST_LOG=debug RUST_BACKTRACE=1 cargo run -- test --chip STM32G031K8Tx --probe 0483:374B
   Compiling probe-rs v0.29.0 (/Users/bmatt/Development/probe-rs/probe-rs)
   Compiling smoke_tester v0.29.0 (/Users/bmatt/Development/probe-rs/smoke-tester)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 3.11s
     Running `/Users/bmatt/Development/probe-rs/target/debug/smoke_tester test --chip STM32G031K8Tx --probe '0483:374B'`
2025-09-19T19:11:16.043390Z DEBUG probe_rs::config::registry: Searching registry for chip with name STM32G031K8Tx
2025-09-19T19:11:16.044173Z DEBUG probe_rs::config::registry: Exact match for chip name: STM32G031K8Tx
2025-09-19T19:11:16.045435Z  INFO probe_rs::config::target: Using sequence Arm(Stm32Armv6 { family: G0 })
[1/1](STM32G031K8Tx) - Starting Test
2025-09-19T19:11:16.048034Z DEBUG nusb::platform::macos_iokit::enumeration: Probing device 100536e5b
```

thenewwazoo commented:
I see that the automated DUT tests are failing, but unfortunately I can't repro the failures.

I can't quite figure out what probes and boards are being tested (especially in the case of [this run](https://github.com/probe-rs/probe-rs/actions/runs/17867713266/job/50814422860?pr=3562), which is hard to parse). Is the test setup documented anywhere?

bugadani commented:
The probes are a mixture of FTDI 232-like devices, ST-links, j-links and ESP32 USB Serial/JTAG peripherals. The MCUs are various ESP32 chips, RP2040, STM and nRF microcontrollers. I don't think my setup has any debugprobes (the one you test with), perhaps the other tester has one.

thenewwazoo commented:
The RPi Pico Probe is "just" a CMSIS-DAP device, so I expect it exercises that probe code.

I've also successfully tested with an ST-Link v2-1 + STM32G031K8Tx (Nucleo-G031K8)
```
$ cargo run -- test --chip STM32G031K8Tx
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.25s
     Running `/Users/bmatt/Development/probe-rs/target/debug/smoke_tester test --chip STM32G031K8Tx`
[1/1](STM32G031K8Tx) - Starting Test
[1/1](STM32G031K8Tx) - Probe: "ST-Link V2-1"
[1/1](STM32G031K8Tx) - Chip:  "STM32G031K8Tx"
[1/1](STM32G031K8Tx) - Core 0: Armv6m
[1/1](STM32G031K8Tx) - Halting core..
[1/1](STM32G031K8Tx) - Test [1] - Testing HW breakpoints
[1/1](STM32G031K8Tx) - Test [1] - Testing region: BANK_1 (0x08000000 - 0x08010000)
[1/1](STM32G031K8Tx) - Test [1] - 4 breakpoints supported
[1/1](STM32G031K8Tx) - Test [1] - Test passed in 28 ms.
[1/1](STM32G031K8Tx) - Test [2] - Testing region: SRAM (0x20000000 - 0x20002000)
[1/1](STM32G031K8Tx) - Test [2] - Test - RAM Start 32
[1/1](STM32G031K8Tx) - Test [2] - Test - RAM End 32
[1/1](STM32G031K8Tx) - Test [2] - Test - RAM Start 8
[1/1](STM32G031K8Tx) - Test [2] - Test - RAM 8 Unaligned
[1/1](STM32G031K8Tx) - Test [2] - Test - RAM End 8
[1/1](STM32G031K8Tx) - Test [2] - Testing: write and read at address 0x20000000: 1 byte at RAM start
[1/1](STM32G031K8Tx) - Test [2] - Testing: write and read at address 0x20000000: 4 bytes at RAM start
[1/1](STM32G031K8Tx) - Test [2] - Testing: write and read at address 0x20001FFC: 4 bytes at RAM end
[1/1](STM32G031K8Tx) - Test [2] - Test passed in 5 ms.
[1/1](STM32G031K8Tx) - Test [3] - Testing register write...
[1/1](STM32G031K8Tx) - Test [3] - Test passed in 38 ms.
[1/1](STM32G031K8Tx) - Test [4] - Testing register read...
[1/1](STM32G031K8Tx) - Test [4] - Test passed in 21 ms.
[1/1](STM32G031K8Tx) - Test [5] - Testing stepping on core 0...
[1/1](STM32G031K8Tx) - Test [5] - R0 value ok!
[1/1](STM32G031K8Tx) - Test [5] - Core halted at 0x20000002, now trying to run...
[1/1](STM32G031K8Tx) - Test [5] - Core halted again!
[1/1](STM32G031K8Tx) - Test [5] - Core halted at 0x20000006, now trying to run...
[1/1](STM32G031K8Tx) - Test [5] - Run core again, with pc = 0x20000008
[1/1](STM32G031K8Tx) - Test [5] - Test passed in 39 ms.
[1/1](STM32G031K8Tx) - Test [6] - Missing resource for test: No flash test binary specified
[1/1](STM32G031K8Tx) - Tests Passed
[DONE] - All DUTs passed.

Test summary:
 [STM32G031K8Tx] passed
All tests passed.
```

I've got a Black Magic Probe and a J-Link on-hand, I just need to get them set up and talking to a real device (validating them before-and-after my change).

bugadani commented:
I suspect the major difference might be macos, I think both of the smoke tester setups run on a raspberry pi with some linux-based image. Your successes are at the very least promising :)

bugadani commented:
To be fair the current failures don't look too bad, it could be that a past failure put the devices in a wrong state.

kevinmehall commented on probe-rs/src/probe/usb_util.rs:
```diff
@@ -10,33 +11,72 @@ pub trait InterfaceExt {
-            comp.status.map_err(io::Error::other)?;
+        let mut endpoint = self
+            .endpoint::<Bulk, Out>(endpoint)
+            .map_err(io::Error::from)?;
```
Instead of creating the endpoint for every transfer, consider constructing it on initialization and passing it around instead of the `Interface` + endpoint address. Also consider whether the [high-level buffered API](https://docs.rs/nusb/latest/nusb/io/index.html) could replace these `read_bulk` / `write_bulk` methods with `read_exact` and `write_all` + `flush`.

thenewwazoo replied:
I considered creating an endpoint and storing it, but since InterfaceExt is an extension of `nusb::Interface`, there's nowhere to store/cache the created endpoint. I could possibly create some kind of global cache but that feels heavy-handed. Ultimately, I think, this would be good work for a future refactor, but I'm trying super hard not to have to make big sweeping changes to all the probe implementations and instead prefer something closer to drop-in. LMK if that's not ambitious enough :)

Fortunately, I think creating an endpoint is a pretty cheap operation - it's mostly afaict just gathering some members and doing a synchronized claim, so there's memory churn and maybe some spinning if someone is doing something heavy with multiple endpoints at the same time, but my gut says that's fairly rare.

Again, totally open to making deeper changes, I just don't want to do so without maintainer concurrence.

kevinmehall replied:
Yeah, it shouldn't be too expensive to re-create -- it's mostly what v0.1 was doing on every operation. Definitely not worth caching in a hash table by address or something like that. Only if there's a convenient place to put the `Endpoint`, like in place of the fields that currently store the endpoint address, but it makes sense to start with the minimal change.

kevinmehall commented on probe-rs/src/probe/usb_util.rs:
```diff
@@ -10,33 +11,72 @@ pub trait InterfaceExt {
-            Ok(n)
+        if buf.is_empty() {
+            return Ok(0);
+        }
```
An OUT transfer of zero bytes sends a zero-length packet. Don't know if any probes rely on that, but it's observable by the device, sometimes used as a delimiter, and not a no-op.

thenewwazoo replied:
Interesting! I will remove the checks from both write and read in the next commit.

kevinmehall commented on probe-rs/src/probe/usb_util.rs:
```diff
@@ -10,33 +11,72 @@ pub trait InterfaceExt {
+        endpoint.submit(transfer_buffer);
+
+        let Some(completion) = endpoint.wait_next_complete(timeout) else {
+            endpoint.cancel_all();
```
`cancel_all` requests the transfer to complete early with a cancelled status, but it happens asynchronously and this doesn't wait for it to finish. You'll want another `wait_next_complete` here to consume the transfer. When `endpoint` goes out of scope and drops with a pending transfer, it too requests cancellation, but it is internally reference counted and the endpoint does not become available to claim again until the cancellation finishes and releases it.

This is a potential cause of the `endpoint already in use` you're hitting in CI -- if there's something handling a timeout error and then immediately calling `read_bulk`/`write_bulk` again, you'll race the transfer cancellation for whether you can claim the endpoint again.

kevinmehall replied:
I should perhaps add a convenience method for submitting and waiting for a single transfer with blocking timeout like this. This API got a lot lower-level in v0.2 because the v0.1 async API was a footgun regarding cancel-safety, but those concerns mostly don't apply to blocking. I think probe-rs wants to switch to async because it will be required for WebUSB support, though.

thenewwazoo replied:
I've added a step to wait for the completion after a cancel, but I'm not sure if it's necessary to explicitly surface the failure of a cancellation to complete in time so the user can opt to reset something farther up the stack. FWIW, I don't see a clean way to even differentiate that based on the error types on-hand.

kevinmehall replied:
The `while let Some(cancelled)` doesn't seem right. It's possible for the transfer to return with a successful status there if the transfer completed normally after the previous `wait_next_complete` timed out but before `cancel_all` reached the host controller. In that case, this would go around the loop and `wait_next_complete` would panic with no transfer pending. 

If you're OK just dropping the returned data and returning TimedOut in that rare case, you could just ignore the result of this call.

thenewwazoo replied:
Good call - I spent some time digging around and thinking, and my big concern was what happens if:
* transfer issued
* timeout
* cancel_all called
* transfer result arrives
* ???

I'm preeeeeetty sure the answer is "the stack throws it away and we never find out" but dude I am a _newbie_ to USB :)

thenewwazoo reviewed (commented):
Thank you for the review @kevinmehall!

thenewwazoo committed:
Wait for cancellation to clear, and actually send ZLPs

Per review comments, this change will not exit early when attempts are
made to write or read zero-length buffers.

This change also makes the reader and writer wait for cancellations to
complete.  Cancellations are performed asynchronously, and keep the
endpoint claim alive; This causes situations where the next read/write
will race with the cancellation.

thenewwazoo commented:
Smoke tests pass on a RP235x+Pico Probe and STM32G070RB+ST-Link v2.1 setup with my most recent commit.

thenewwazoo commented:
(please correct me if I shouldn't record smoke test runs here)

Success smoke-testing an RP2040 connected to a SEGGER Jlink:

```
$ cargo run -- test  --probe 1366:0101 --chip rp2040
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.19s
     Running `/Users/bmatt/Development/probe-rs/target/debug/smoke_tester test --probe '1366:0101' --chip rp2040`
[1/1](RP2040) - Starting Test
[1/1](RP2040) - Probe: "J-Link"
[1/1](RP2040) - Chip:  "RP2040"
[1/1](RP2040) - Core 0: Armv6m
[1/1](RP2040) - Halting core..
[1/1](RP2040) - Test [1] - Testing HW breakpoints
[1/1](RP2040) - Test [1] - Testing region: <unnamed region> (0x10000000 - 0x18000000)
[1/1](RP2040) - Test [1] - 4 breakpoints supported
[1/1](RP2040) - Test [1] - Test passed in 60 ms.
[1/1](RP2040) - Test [2] - Testing region: <unnamed region> (0x20000000 - 0x20042000)
[1/1](RP2040) - Test [2] - Test - RAM Start 32
[1/1](RP2040) - Test [2] - Test - RAM End 32
[1/1](RP2040) - Test [2] - Test - RAM Start 8
[1/1](RP2040) - Test [2] - Test - RAM 8 Unaligned
[1/1](RP2040) - Test [2] - Test - RAM End 8
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x20000000: 1 byte at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x20000000: 4 bytes at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x20041FFC: 4 bytes at RAM end
[1/1](RP2040) - Test [2] - Testing region: <unnamed region> (0x21000000 - 0x21010000)
[1/1](RP2040) - Test [2] - Test - RAM Start 32
[1/1](RP2040) - Test [2] - Test - RAM End 32
[1/1](RP2040) - Test [2] - Test - RAM Start 8
[1/1](RP2040) - Test [2] - Test - RAM 8 Unaligned
[1/1](RP2040) - Test [2] - Test - RAM End 8
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21000000: 1 byte at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21000000: 4 bytes at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x2100FFFC: 4 bytes at RAM end
[1/1](RP2040) - Test [2] - Testing region: <unnamed region> (0x21010000 - 0x21020000)
[1/1](RP2040) - Test [2] - Test - RAM Start 32
[1/1](RP2040) - Test [2] - Test - RAM End 32
[1/1](RP2040) - Test [2] - Test - RAM Start 8
[1/1](RP2040) - Test [2] - Test - RAM 8 Unaligned
[1/1](RP2040) - Test [2] - Test - RAM End 8
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21010000: 1 byte at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21010000: 4 bytes at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x2101FFFC: 4 bytes at RAM end
[1/1](RP2040) - Test [2] - Testing region: <unnamed region> (0x21020000 - 0x21030000)
[1/1](RP2040) - Test [2] - Test - RAM Start 32
[1/1](RP2040) - Test [2] - Test - RAM End 32
[1/1](RP2040) - Test [2] - Test - RAM Start 8
[1/1](RP2040) - Test [2] - Test - RAM 8 Unaligned
[1/1](RP2040) - Test [2] - Test - RAM End 8
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21020000: 1 byte at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21020000: 4 bytes at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x2102FFFC: 4 bytes at RAM end
[1/1](RP2040) - Test [2] - Testing region: <unnamed region> (0x21030000 - 0x21040000)
[1/1](RP2040) - Test [2] - Test - RAM Start 32
[1/1](RP2040) - Test [2] - Test - RAM End 32
[1/1](RP2040) - Test [2] - Test - RAM Start 8
[1/1](RP2040) - Test [2] - Test - RAM 8 Unaligned
[1/1](RP2040) - Test [2] - Test - RAM End 8
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21030000: 1 byte at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x21030000: 4 bytes at RAM start
[1/1](RP2040) - Test [2] - Testing: write and read at address 0x2103FFFC: 4 bytes at RAM end
[1/1](RP2040) - Test [2] - Test passed in 69 ms.
[1/1](RP2040) - Test [3] - Testing register write...
[1/1](RP2040) - Test [3] - Test passed in 88 ms.
[1/1](RP2040) - Test [4] - Testing register read...
[1/1](RP2040) - Test [4] - Test passed in 48 ms.
[1/1](RP2040) - Test [5] - Testing stepping on core 0...
[1/1](RP2040) - Test [5] - R0 value ok!
[1/1](RP2040) - Test [5] - Core halted at 0x20000002, now trying to run...
[1/1](RP2040) - Test [5] - Core halted again!
[1/1](RP2040) - Test [5] - Core halted at 0x20000006, now trying to run...
[1/1](RP2040) - Test [5] - Run core again, with pc = 0x20000008
[1/1](RP2040) - Test [5] - Test passed in 126 ms.
[1/1](RP2040) - Test [6] - Missing resource for test: No flash test binary specified
[1/1](RP2040) - Tests Passed
[DONE] - All DUTs passed.

Test summary:
 [RP2040] passed
All tests passed.
```

I _believe_ that this setup reproduces [this failing test in CI](https://github.com/probe-rs/probe-rs/actions/runs/17882961799/job/50853126054?pr=3562#step:3:453):

```
[12/14](RP2040) - Starting Test - /home/probe-rs-runner/dut_definitions/rp2040.toml
[12/14](RP2040) - Error message: Failed to open probe with selector 1366:0101:000000123456: An error which is specific to the debug probe in use occurred.: A USB transport error occurred.: bulk write timed out
[12/14](RP2040) -  caused by:    An error which is specific to the debug probe in use occurred.
[12/14](RP2040) - Tests Failed
```

thenewwazoo commented:
Well, that was easy! ESP32S3 is _super_ simple to set up and passed the smoke-test:

```
✔ ~/Development/probe-rs/smoke-tester [bmatt/nusb-0.2|…2]$ cargo run -- test  --probe 303a:1001
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.35s
     Running `/Users/bmatt/Development/probe-rs/target/debug/smoke_tester test --probe '303a:1001'`

thread 'main' panicked at smoke-tester/src/main.rs:106:40:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
✘-101 ~/Development/probe-rs/smoke-tester [bmatt/nusb-0.2|…2]$ cargo run -- test  --probe 303a:1001 --chip esp32s3
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.24s
     Running `/Users/bmatt/Development/probe-rs/target/debug/smoke_tester test --probe '303a:1001' --chip esp32s3`
[1/1](esp32s3) - Starting Test
[1/1](esp32s3) - Probe: "Esp USB JTAG"
[1/1](esp32s3) - Chip:  "esp32s3"
[1/1](esp32s3) - Core 0: Xtensa
[1/1](esp32s3) - Halting core..
[1/1](esp32s3) - Test [1] - Testing HW breakpoints
[1/1](esp32s3) - Test [1] - Testing region: <unnamed region> (0x00000000 - 0x04000000)
[1/1](esp32s3) - Test [1] - 2 breakpoints supported
[1/1](esp32s3) - Test [1] - Testing region: External data bus (0x3C000000 - 0x3E000000)
[1/1](esp32s3) - Test [1] - 2 breakpoints supported
[1/1](esp32s3) - Test [1] - Testing region: External instruction bus (0x42000000 - 0x44000000)
[1/1](esp32s3) - Test [1] - 2 breakpoints supported
[1/1](esp32s3) - Test [1] - Test passed in 9 ms.
[1/1](esp32s3) - Test [2] - Testing region: SRAM1 Data bus (0x3FC88000 - 0x3FCF0000)
[1/1](esp32s3) - Test [2] - Test - RAM Start 32
[1/1](esp32s3) - Test [2] - Test - RAM End 32
[1/1](esp32s3) - Test [2] - Test - RAM Start 8
[1/1](esp32s3) - Test [2] - Test - RAM 8 Unaligned
[1/1](esp32s3) - Test [2] - Test - RAM End 8
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x3FC88000: 1 byte at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x3FC88000: 4 bytes at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x3FCEFFFC: 4 bytes at RAM end
[1/1](esp32s3) - Test [2] - Testing region: SRAM2 Data bus (0x3FCF0000 - 0x3FD00000)
[1/1](esp32s3) - Test [2] - Test - RAM Start 32
[1/1](esp32s3) - Test [2] - Test - RAM End 32
[1/1](esp32s3) - Test [2] - Test - RAM Start 8
[1/1](esp32s3) - Test [2] - Test - RAM 8 Unaligned
[1/1](esp32s3) - Test [2] - Test - RAM End 8
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x3FCF0000: 1 byte at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x3FCF0000: 4 bytes at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x3FCFFFFC: 4 bytes at RAM end
[1/1](esp32s3) - Test [2] - Testing region: SRAM1 Instruction bus (0x40370000 - 0x40378000)
[1/1](esp32s3) - Test [2] - Test - RAM Start 32
[1/1](esp32s3) - Test [2] - Test - RAM End 32
[1/1](esp32s3) - Test [2] - Test - RAM Start 8
[1/1](esp32s3) - Test [2] - Test - RAM 8 Unaligned
[1/1](esp32s3) - Test [2] - Test - RAM End 8
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x40370000: 1 byte at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x40370000: 4 bytes at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x40377FFC: 4 bytes at RAM end
[1/1](esp32s3) - Test [2] - Testing region: SRAM2 Instruction bus (0x40378000 - 0x403E0000)
[1/1](esp32s3) - Test [2] - Test - RAM Start 32
[1/1](esp32s3) - Test [2] - Test - RAM End 32
[1/1](esp32s3) - Test [2] - Test - RAM Start 8
[1/1](esp32s3) - Test [2] - Test - RAM 8 Unaligned
[1/1](esp32s3) - Test [2] - Test - RAM End 8
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x40378000: 1 byte at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x40378000: 4 bytes at RAM start
[1/1](esp32s3) - Test [2] - Testing: write and read at address 0x403DFFFC: 4 bytes at RAM end
[1/1](esp32s3) - Test [2] - Test passed in 27 ms.
[1/1](esp32s3) - Test [3] - Testing register write...
[1/1](esp32s3) - Test [3] - Test passed in 3 ms.
[1/1](esp32s3) - Test [4] - Testing register read...
[1/1](esp32s3) - Test [4] - Test passed in 0 ms.
[1/1](esp32s3) - Test [5] - Testing stepping on core 0...
[1/1](esp32s3) - Test [5] - Test not implemented for esp32s3: Testing stepping is not implemented for Xtensa yet.
[1/1](esp32s3) - Test [6] - Missing resource for test: No flash test binary specified
[1/1](esp32s3) - Tests Passed
[DONE] - All DUTs passed.

Test summary:
 [esp32s3] passed
All tests passed.
```

I believe that covers [this failed test in CI](https://github.com/probe-rs/probe-rs/actions/runs/17882961799/job/50853126050?pr=3562#step:3:81) (but again I'm not 100% sure, that log is a bit hard to suss out).

thenewwazoo commented:
I dug around and don't have an MPSSE-capable FT?232[HD] board around, so I'm not going to be able to test those.

I do have a Black Magic Probe and have gotten it to tell me it doesn't have multi-drop to talk to an RP235x 😆. I'm going to try to get it talking to another chip soon and see if I can get it to smoke-test.

thenewwazoo committed:
Ignore result of cancellation after a timeout

bugadani commented:
Well the tests are green, but the embedded-test output has more USB errors than I'd like. Not sure if this is just a transitive error or not, or whether it will be easy to debug...

thenewwazoo commented:
> Well the tests are green, but the embedded-test output has more USB errors than I'd like. Not sure if this is just a transitive error or not, or whether it will be easy to debug...

I've got an RPi lying around, along with the various probes I've mentioned testing with my macOS laptop. If you can provide more info about the various test configurations, I can try to replicate them to cover the specific failures.

thenewwazoo commented:
Oh, I was more asking about the particular probes - I have been able to make some guesses based on VIDs and PIDs, but it'd be nice to have a list.

bugadani commented:
I don't have the exact list with me but this is what I remember or can piece together:
- One of the STM32F050 is a discovery kit, reflashed with SEGGER's old st-link -> j-link converter.
- RP2040 is a clone j-link like [this one](https://www.aliexpress.com/p/tesla-landing/index.html?scenario=c_ppc_item_bridge&productId=1005005994095533).
- nrf51/52 are nordic devboards with on-board jlinks
- ESP32C2, ESP32 and STM32F429 are FTDI FT2232H chips in the form of esp-prog boards.
- ESP32S3, C3 and C6 are ESP USB JTAG devices
- The rest are various STM discovery or nucleo boards, unmodified ST-links of various versions.

The smoke-tester2 and embedded-test devices are the same, even if the set of tested devices differs slightly for reasons.

thenewwazoo committed:
Update Cargo.lock after rebasing on master

thenewwazoo commented:
I spent some time today building examples and testing probes attached to an RPi 4B I had lying around (prior tests were performed on my macOS laptop). I've [attached (boards.zip)](https://github.com/user-attachments/files/22628753/boards.zip) the DUT files, the binaries I used to flash, and the logs from the smoke-tester. The combos I tested:

* SEGGER JLink + RP2040 (SWD)
* Nucleo-G070RB w/ onboard ST-Link v2-1
* ESP32S3 w/ onboard USB-attached JTAG
* Raspberry Pi Pico Probe (i.e. CMSIS-DAP) + RP2350
* Black Magic Probe + ATSAMD21G18AU (i.e. Feather M0)

The only problem of note was flashing the ATSAMD w/ the BMP, but interestingly that fails w/ the tip of master too, so at least that isn't a regression. In every other case, the combo smoke-tested and ran without issue.

I bought a FT232HQ breakout, but have nothing to attach it to - every device I have on-hand exposes SWD.

It's worth noting that I have also successfully used this PR's code to run Humility to flash and attach to an RP235x on both macOS and Illumos.

At this point, I'm not sure what more I can do to test, but would appreciate any guidance. For my part, I'm pretty confident in the code! :)

bugadani commented:
I can reproduce issues on a selection of my hardware:

STM32F051R8 discovery board with a reflashed J-Link programmer:

```
probe-rs run .\stm32f051r8 --chip stm32f051r8
      Erasing ✔ 100% [####################]  17.00 KiB @  14.58 KiB/s (took 1s)
  Programming ✔ 100% [####################]  17.00 KiB @  11.35 KiB/s (ETA 0s)                                           WARN probe_rs::session: Could not clear all hardware breakpoints: An ARM specific error occurred.

Caused by:
    0: The debug probe encountered an error.
    1: An error which is specific to the debug probe in use occurred.
    2: A USB transport error occurred.
    3: bulk write timed out
 WARN probe_rs::session: Failed to deconfigure device during shutdown: Arm(Probe(ProbeSpecific(BoxedProbeError(Usb(Custom { kind: TimedOut, error: "bulk write timed out" })))))
Error: An error with the flashing procedure has occurred.

Caused by:
    0: The uninitialization of the flash algorithm failed.
    1: Failed to read the core status.
    2: An ARM specific error occurred.
    3: Error using access port FullyQualifiedApAddress { dp: Default, ap: V1(0) }.
    4: Failed to read register DRW at address 0xd0c
    5: The debug probe encountered an error.
    6: An error which is specific to the debug probe in use occurred.
    7: A USB transport error occurred.
    8: endpoint already in use
```

ESP32 (ethernet devkit, FT2232H):

```
 WARN probe_rs::vendor::espressif::sequences::esp32: Be careful not to reset your ESP32 while connected to the debugger! Depending on the specific device, this may render it temporarily inoperable or permanently damage it.
 WARN probe_rs::probe::ftdi: Read 1 bytes, expected 4
 WARN probe_rs::session: Could not clear all hardware breakpoints: An error with the usage of the probe occurred

Caused by:
    A timeout occurred during probe operation.
 WARN probe_rs::probe::ftdi: Read 1 bytes, expected 5
 WARN probe_rs::session: Failed to deconfigure device during shutdown: Probe(Timeout)
Error: Connecting to the chip was unsuccessful.

Caused by:
    0: An error with the usage of the probe occurred
    1: An error which is specific to the debug probe in use occurred.
    2: A USB transport error occurred.
    3: endpoint already in use
```

ESP32-C2 via FT232H:

```
 WARN probe_rs::probe::ftdi: Read 6 bytes, expected 12
 WARN probe_rs::session: Could not clear all hardware breakpoints: An error with the usage of the probe occurred

Caused by:
    A timeout occurred during probe operation.
 WARN probe_rs::probe::ftdi: Read 6 bytes, expected 18
 WARN probe_rs::session: Failed to deconfigure device during shutdown: Probe(Timeout)
Error: The uninitialization of the flash algorithm failed.

Caused by:
    0: Failed to read the core status.
    1: An error with the usage of the probe occurred
    2: An error which is specific to the debug probe in use occurred.
    3: A USB transport error occurred.
    4: endpoint already in use
error: process didn't exit successfully: `probe-rs run --preverify target\riscv32imc-unknown-none-elf\release\esp_radio_init` (exit code: 1)
```

The common thread in all this is the `endpoint already in use` error and that the FTDI driver reads unexpected amounts of data.

thenewwazoo commented:
I'm running tests on both macOS and an RPi 4B running Bookworm. I'll keep poking at it - I noticed you're using `probe-rs run` and not just `smoke-test` so I'm going to try to repro the failures that way (as soon as it finishes building... which takes forever on this Pi! :) )

thenewwazoo commented:
@bugadani thank you for having so much patience while I try to replicate the issue. Is there any chance you could attach a debug-level log of a failed test run?

kevinmehall commented:
I was able to reproduce similar "endpoint already in use" errors in my testing. It's a bug in nusb, triggered by quickly [dropping and re-creating](https://github.com/probe-rs/probe-rs/pull/3562#discussion_r2364991165) an `Endpoint` shortly after a previous transfer completes. It's a race when the events thread hasn't let go of `Endpoint`'s internal `Arc` yet so it's still considered in use. I'll get you a fix soon.

kevinmehall commented:
Try [nusb v0.2.1](https://github.com/kevinmehall/nusb/releases/tag/v0.2.1). I also added `Endpoint::transfer_blocking` that could simplify the timeout handling in `InterfaceExt`.

thenewwazoo committed:
Bump to nusb 0.2.1 for race condition fix

thenewwazoo committed:
Update lockfile

thenewwazoo commented:
> @thenewwazoo any news on this one?

Sadly, no. I've run the smoke tester several times on my RPi connected to the esp32s3 I have on my bench, and the only problem I've noticed is that the chip appears to be fragile when interrupting operations. I sometimes have to reset the bus/power-cycle the device when I kill the smoke-tester process mid-flash.

I'm currently running the smoker-tester in a loop to torture it. We'll see what happens. AFAICT that's the only device remaining in the CI test suite that's failing, and only on one host.

bugadani reviewed (approved):
Seems to be working fine now, thank you, and I also thank you for your rigorous torture-testing on your S3 :) Well done
